### PR TITLE
Adds affiliations to work form.

### LINKED ISSUE #3156: Add user interface for multivalued affiliation field next for each contributor
Mapping work that goes along with this in #3099

Design: https://projects.invisionapp.com/share/NP138M0QX7T3#/screens/473166568
Requirements: https://docs.google.com/presentation/d/1FI9yn7vUSeX7Yd5IemrLxnYH2__JlTxNwgP8guATb18

_Background:_
- we want to record contributor affiliations along with names (and ORCIDs)
- recording affiliations, especially with organization identifiers (RoRs) will enrich metadata and discoverability

_Requirement:_
- Contributor entries on the work edit form should include a new affiliation entry box.
- The field is multivalued so needs an "Add another affiliation" button.
- The value(s) are associated with the contributor.

_Notes:_
- Affiliation is not required
- Multiple affiliations are allowed
- Free text entries are allowed; ROR not required but preferred (see [#3100](https://github.com/sul-dlss/happy-heron/issues/3100) for RoR)

There is a separate ticket for how/what we need to store and how it maps to cocina (in #3099).  If questions about what we need to store are needed before the cocina mapping work is done, ask if Arcadia can quickly answer to unblock work.

### PR DESCRIPTION
**Pull request by justinlittman**
closes #3156

# Why was this change made? 🤔

# How was this change tested? 🤨

⚡ ⚠ If this change involves consuming from or writing to another service (or shared file system), ***run [integration test create_object_h2_spec.rb](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test manually in [stage|qa] environment, in addition to specs. ⚡

Unit, Amy

# Does your change introduce accessibility violations? 🩺

⚡ ⚠ Please ensure this change does not introduce accessibility violations (at the WCAG A or AA conformance levels); if it does, include a rationale. See the [Infrastructure accessibility guide](https://github.com/sul-dlss/DeveloperPlaybook/blob/main/best-practices/infra-accessibility.md) for more detail. ⚡

No, per SiteImprove.

### CODE REVIEW

**Comment by justinlittman** [comment]
image

File: app/components/works/form_component.html.erb [hunk 1]
```diff
@@ -9,11 +9,11 @@
       html: {class: "needs-validation work-editor", novalidate: true, multipart: true} do |f| %>
   <% if work_form.errors.present? %>
     <div id="error_explanation" class="alert alert-danger">
-      <h4><%= pluralize(work_form.errors.count, "error") %> prohibited this item from being saved:</h4>
+      <h4><%= pluralize(work_form.errors.size, "error") %> prohibited this item from being saved:</h4>
```
Comment (justinlittman) [inline app/components/works/form_component.html.erb:12]: Needed to make some reform-related tweaks to get the errors to display correctly.

File: app/components/works/affiliations_row_component.html.erb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        <%= form.hidden_field :uri, {"data-autocomplete-target": "hidden", "aria-label": "URI for affiliation", class: "affiliation-input"} %>
+        <ul class="list-group" data-autocomplete-target="results"></ul>
+      </div>
+  </div>
+  <div class="col-sm-5">
+      <%= form.label :department, "Department/Institute/Center", class: "col-form-label" %>
+      <%= form.text_field :department, class: "form-control affiliation-input" %>
+  </div>
+  <div class="offset-sm-1 col-sm-1">
+    <%= button_tag type: "button", class: "btn btn-sm", aria: {label: "Remove"},
```
Comment (mjgiarlo) [inline app/components/works/affiliations_row_component.html.erb:17]: Wondering if the aria label could make it clearer what's being removed, e.g., `Remove affiliation`?

File: app/components/works/affiliations_row_component.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  # Add a single affiliation to an author / contributor.
+  class AffiliationsRowComponent < ApplicationComponent
+    def initialize(form:)
+      @form = form
+    end
+
+    attr_reader :form
+
+    def error?
+      errors.present?
```
Comment (mjgiarlo) [inline app/components/works/affiliations_row_component.rb:13]: I don't see this called anywhere explicitly. Wired up automagically elsewhere?

File: app/components/works/show/authors_contributors_component.html.erb [hunk 4]
```diff
@@ -5,13 +5,16 @@
...
             <%= participant.first_name %> <%= participant.last_name %>
-          </td>
-        <% else %>
-          <td style="width: 50%"><%= participant.full_name %></td>
-        <% end %>
+          <% else %>
+            <%= participant.full_name %>
+          <% end %>
+          <% if participant.affiliations.any? %>
+            (<%= participant.affiliations.map { |aff| aff.department ? "#{aff.label}, #{aff.department}" : aff.label }.join("; ") %>)
```
Comment (mjgiarlo) [inline app/components/works/show/authors_contributors_component.html.erb:15]: I wonder if we should consider adding this affiliation display logic to the model. So here we'd have:

```suggestion
            (<%= participant.affiliations.map(&:to_s).join("; ") %>)
```

And in the `Affiliation` model we'd have something like:

```ruby
class Affiliation < ApplicationRecord
  # ...
  def to_s
    return label if department.blank?

    "#{label}, #{department}"
  end
  # or the equivalent:
  def to_s
    [label, department].compact_blank.join(", ")
  end
  # ...
end
```

File: db/structure.sql [hunk 5]
```diff
@@ -297,7 +297,10 @@ CREATE TABLE public.collections (
     review_enabled boolean DEFAULT false,
     license_option character varying DEFAULT 'required'::character varying NOT NULL,
     head_id bigint,
-    doi_option character varying DEFAULT 'yes'::character varying
+    doi_option character varying DEFAULT 'yes'::character varying,
+    allow_custom_rights_statement boolean DEFAULT false NOT NULL,
+    provided_custom_rights_statement character varying,
+    custom_rights_statement_custom_instructions character varying
```
Comment (mjgiarlo) [inline db/structure.sql:303]: Is this part of the affiliations work?

File: app/components/works/affiliations_row_component.html.erb (hunk 2 above)
Comment (justinlittman) [reply app/components/works/affiliations_row_component.html.erb:17]: Agree. Fixed.

File: app/components/works/show/authors_contributors_component.html.erb (hunk 4 above)
Comment (justinlittman) [reply app/components/works/show/authors_contributors_component.html.erb:15]: Much cleaner. Fixed.

File: app/components/works/affiliations_row_component.rb (hunk 3 above)
Comment (justinlittman) [reply app/components/works/affiliations_row_component.rb:13]: Copypasta. Removed.

File: db/structure.sql (hunk 5 above)
Comment (justinlittman) [reply db/structure.sql]: No. Removed.

**Review by mjgiarlo — APPROVED** [review]

File: app/components/works/show/authors_contributors_component.html.erb (hunk 4 above)
Comment (mjgiarlo) [reply app/components/works/show/authors_contributors_component.html.erb]: Thanks!

**Pull request merged**